core/capabilities/compute/compute.go

```diff
@@ -0,0 +1,202 @@
+package compute
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	capabilitiespb "github.com/smartcontractkit/chainlink-common/pkg/capabilities/pb"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	coretypes "github.com/smartcontractkit/chainlink-common/pkg/types/core"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/host"
+	wasmpb "github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/pb"
+)
+
+const (
+	CapabilityIDCompute = "custom_compute@1.0.0"
+
+	binaryKey = "binary"
+	configKey = "config"
+)
+
+var (
+	computeWASMInit = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_init",
+		Help: "how long it takes to initialize a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+	computeWASMExec = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_exec",
+		Help: "how long it takes to execute a request from a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+)
+
+var _ capabilities.ActionCapability = (*Compute)(nil)
+
+type Compute struct {
+	log      logger.Logger
+	registry coretypes.CapabilitiesRegistry
+	modules  *moduleCache
+}
+
+func (c *Compute) RegisterToWorkflow(ctx context.Context, request capabilities.RegisterToWorkflowRequest) error {
+	return nil
+}
+
+func (c *Compute) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+func generateID(binary []byte) string {
+	id := sha256.Sum256(binary)
+	return fmt.Sprintf("%x", id)
+}
+
+func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	binary, err := c.popBytesValue(request.Config, binaryKey)
```

Comment:
Config is static to the workflow right?
Would it make sense to do lines 84-94 in `RegisterToWorkflow` and hold in state rather than on every Excecute?

Reply:
It is, but as I say below I'm trying to avoid writing a stateful capability wherever possible

Come to think of it, this should probably not mutate the incoming config, so let me address that